### ISSUE TITLE
How to combine two frozen models (Tensorflow) for object detection?

### ISSUE THREAD

**Issue by balavenkatesh3322** [tensorflow/tensorflow#18080]
I am trying to combine two frozen models (protobuffs) for object detection. The issue is one of the models is my own dataset and the other is the prebuilt model for coco dataset (just include more classes to the dataset itself). Is this possible? or is there a better approach to perform this?

**Comment by jinay1991** [comment]
For object detection why you even need to combine models, where you can use both models and get best of both.

Run model-1 on input image and run model-2 on input image.
Draw boxes for result of model-1 and result of model-2 on same image will give you detections for both COCO classes and your own classes.

Another thing you can do is to fine-tune COCO model with new dataset to add new classes. See tranfer Learning to know how to do that.

**Comment by balavenkatesh3322** [comment]
How it will read two model file at the same time in tensorflow session.please send me some example code

**Comment by jinay1991** [comment]
```
with tf.gfile.GFile(model_filename1, "rb") as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Graph().as_default() as graph1:
    tf.import_graph_def(graph_def)

with tf.gfile.GFile(model_filename2, "rb") as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())

with tf.Graph().as_default() as graph2:
    tf.import_graph_def(graph_def)

sess1 = tf.Session(graph=graph1)
sess2 = tf.Session(graph=graph2)
```

Above code might not work as it is, as I have not got chance to test these. But idea was that we can create two session type object for each models and call their run methods, obviously if you are using CUDA Hardware you will be blocked by calling run() if both called same time due to locking mechanism of TF. But you can run this by setting Queue or calling run() one after another.

I have loaded two graphs on same GPU (one graph was for preprocessing and another one was object classification) and it worked well for me. (I called run() one after another as I required output of 1st graph to be fed in to 2nd).

Hope you get idea.

**Comment by balavenkatesh3322** [comment]
Below is my code.please look into that and give some clarification about two model loading

```
`# What model to download.
MODEL_NAME = 'ssd_mobilenet_v1_coco_11_06_2017'
MODEL_FILE = MODEL_NAME + '.tar.gz'
DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

# Path to frozen detection graph. This is the actual model that is used for the object detection.
#PATH_TO_CKPT = MODEL_NAME + '/frozen_inference_graph.pb'
PATH_TO_CKPT = 'screwdriver_graph/frozen_inference_graph.pb'


# List of the strings that is used to add correct label for each box.
#PATH_TO_LABELS = os.path.join('data', 'mscoco_label_map.pbtxt')
PATH_TO_LABELS = os.path.join('data', 'screwdriver_label.pbtxt')

NUM_CLASSES = 1
#NUM_CLASSES = 90


# ## Download Model

#if not os.path.exists(MODEL_NAME + '/frozen_inference_graph.pb'):
if not os.path.exists('screwdriver_graph/frozen_inference_graph.pb'):
	print ('Downloading the model')
	opener = urllib.request.URLopener()
	opener.retrieve(DOWNLOAD_BASE + MODEL_FILE, MODEL_FILE)
	tar_file = tarfile.open(MODEL_FILE)
	for file in tar_file.getmembers():
	  file_name = os.path.basename(file.name)
	  if 'frozen_inference_graph.pb' in file_name:
	    tar_file.extract(file, os.getcwd())
	print ('Download complete')
else:
	print ('Model already exists')

# ## Load a (frozen) Tensorflow model into memory.

detection_graph = tf.Graph()
with detection_graph.as_default():
  od_graph_def = tf.GraphDef()
  with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
    serialized_graph = fid.read()
    od_graph_def.ParseFromString(serialized_graph)
    tf.import_graph_def(od_graph_def, name='')


# ## Loading label map
# Label maps map indices to category names, so that when our convolution network predicts `5`, we know that this corresponds to `airplane`.  Here we use internal utility functions, but anything that returns a dictionary mapping integers to appropriate string labels would be fine

label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
category_index = label_map_util.create_category_index(categories)

#intializing the web camera device

import cv2
cap = cv2.VideoCapture(1)



# Running the tensorflow session
with detection_graph.as_default():
  with tf.Session(graph=detection_graph) as sess:
   ret = True
   while (ret):
      ret,image_np = cap.read()
      # Expand dimensions since the model expects images to have shape: [1, None, None, 3]
      image_np_expanded = np.expand_dims(image_np, axis=0)
      image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
      # Each box represents a part of the image where a particular object was detected.
      boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
      # Each score represent how level of confidence for each of the objects.
      # Score is shown on the result image, together with the class label.
      scores = detection_graph.get_tensor_by_name('detection_scores:0')
      classes = detection_graph.get_tensor_by_name('detection_classes:0')
      num_detections = detection_graph.get_tensor_by_name('num_detections:0')
      # Actual detection.
      (boxes, scores, classes, num_detections) = sess.run(
          [boxes, scores, classes, num_detections],
          feed_dict={image_tensor: image_np_expanded})
      # Visualization of the results of a detection.
      vis_util.visualize_boxes_and_labels_on_image_array(
          image_np,
          np.squeeze(boxes),
          np.squeeze(classes).astype(np.int32),
          np.squeeze(scores),
          category_index,
          use_normalized_coordinates=True,
          line_thickness=8)
#      plt.figure(figsize=IMAGE_SIZE)
#      plt.imshow(image_np)
      #cv2.putText(img,'OpenCV',(10,500), font, 4,(255,255,255),2,2)
      cv2.imshow('Detected Tools',cv2.resize(image_np,(1000,850)))
      #cv2.imshow('Tools Name',cv2.resize(cv2.putText(img,'OpenCV',(10,500), font, 4,(255,255,255),2,2),(100,150)))
      if cv2.waitKey(25) & 0xFF == ord('q'):
          cv2.destroyAllWindows()
          cap.release()
          break


`
```

Thanks for helping

**Comment by jinay1991** [comment]
```
import cv2
import tensorflow as tf

# What model to download.
MODEL_NAME = 'ssd_mobilenet_v1_coco_11_06_2017'
MODEL_FILE = MODEL_NAME + '.tar.gz'
DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

# Path to frozen detection graph. This is the actual model that is used for the object detection.
# PATH_TO_CKPT = MODEL_NAME + '/frozen_inference_graph.pb'
PATH_TO_CKPT = 'screwdriver_graph/frozen_inference_graph.pb'


# List of the strings that is used to add correct label for each box.
# PATH_TO_LABELS = os.path.join('data', 'mscoco_label_map.pbtxt')
PATH_TO_LABELS = os.path.join('data', 'screwdriver_label.pbtxt')

NUM_CLASSES = 1
# NUM_CLASSES = 90


# ## Download Model

# if not os.path.exists(MODEL_NAME + '/frozen_inference_graph.pb'):
if not os.path.exists('screwdriver_graph/frozen_inference_graph.pb'):
    print('Downloading the model')
    opener = urllib.request.URLopener()
    opener.retrieve(DOWNLOAD_BASE + MODEL_FILE, MODEL_FILE)
    tar_file = tarfile.open(MODEL_FILE)
    for file in tar_file.getmembers():
        file_name = os.path.basename(file.name)
        if 'frozen_inference_graph.pb' in file_name:
            tar_file.extract(file, os.getcwd())
    print('Download complete')
else:
    print('Model already exists')

# ## Load a (frozen) Tensorflow model into memory.

detection_graph = tf.Graph()
with detection_graph.as_default():
    od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
        serialized_graph = fid.read()
        od_graph_def.ParseFromString(serialized_graph)
        tf.import_graph_def(od_graph_def, name='')

# ---- Load SSD Model as graph_ssd ----
with tf.Graph().as_default() as graph_ssd:
    graph_def = tf.GraphDef()
    with tf.gfile.GFile(SSD_MODEL_FILE, 'rb') as fp:
        graph_def.ParseFromString(fp.read())
    tf.import_graph_def(graph_def)

ssd_sess = tf.Session(graph=graph_ssd)

with tf.Graph().as_default() as graph_custom:
    graph_def = tf.GraphDef()
    with tf.gfile.GFile(CUSTOM_MODEL_FILE, 'rb') as fp:
        graph_def.ParseFromString(fp.read())
    tf.import_graph_def(graph_def)

custom_sess = tf.Session(graph=graph_custom)



# ## Loading label map
# Label maps map indices to category names, so that when our convolution network predicts `5`, we know that this corresponds to `airplane`.  Here we use internal utility functions, but anything that returns a dictionary mapping integers to appropriate string labels would be fine

label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
category_index = label_map_util.create_category_index(categories)

# intializing the web camera device


cap = cv2.VideoCapture(1)


# Running the tensorflow session
# with detection_graph.as_default():
#     with tf.Session(graph=detection_graph) as sess:
ret = True
while (ret):
    ret, image_np = cap.read()
    # Expand dimensions since the model expects images to have shape: [1, None, None, 3]
    image_np_expanded = np.expand_dims(image_np, axis=0)
    image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
    # Each box represents a part of the image where a particular object was detected.
    boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
    # Each score represent how level of confidence for each of the objects.
    # Score is shown on the result image, together with the class label.
    scores = detection_graph.get_tensor_by_name('detection_scores:0')
    classes = detection_graph.get_tensor_by_name('detection_classes:0')
    num_detections = detection_graph.get_tensor_by_name('num_detections:0')
    # Actual detection.
    (boxes, scores, classes, num_detections) = ssd_sess.run(
        [boxes, scores, classes, num_detections],
        feed_dict={image_tensor: image_np_expanded})

    # ----------
    (predictions) = custom_sess.run([classifier_out], feed_dict={image_tensor_custom: image_np_expanded})
    # ----------

    # Visualization of the results of a detection.
    vis_util.visualize_boxes_and_labels_on_image_array(
        image_np,
        np.squeeze(boxes),
        np.squeeze(classes).astype(np.int32),
        np.squeeze(scores),
        category_index,
        use_normalized_coordinates=True,
        line_thickness=8)
    # cv2.putText(img,'OpenCV',(10,500), font, 4,(255,255,255),2,2)
    cv2.imshow('Detected Tools', cv2.resize(image_np, (1000, 850)))
    # cv2.imshow('Tools Name',cv2.resize(cv2.putText(img,'OpenCV',(10,500), font, 4,(255,255,255),2,2),(100,150)))
    if cv2.waitKey(25) & 0xFF == ord('q'):
        cv2.destroyAllWindows()
        cap.release()
        break
```

**Comment by balavenkatesh3322** [comment]
How can we maintain If it will go to multiple model files because of python code will be increased
Is there any other feasible solution?(for loop)

**Comment by aselle** [comment]
@nfiedel, is there a better solution using saved model?

**Comment by nfiedel** [comment]
Possibly, though this will be pretty advanced and likely not work easily out of the box.

TLDR: [TensorFlow Hub](https://www.tensorflow.org/hub/) enables you to compose Hub Modules into a single model. If you can export/provide both models as Hub Modules, then you can later use the Hub APIs to import them into a single graph/model. You will still need to wire-up the inputs & outputs.

**Comment by nfiedel** [comment]
I'm not aware of any sample code for this. Like I mentioned, this is pretty advanced and new area to research / explore.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by SriramganeshMarimuthu** [comment]
@balavenkatesh3322  Did you find any solution for combine two frozen model? if you get any idea please share to me as well.

**Comment by SriramganeshMarimuthu** [comment]
@jinay1991  I checked your code it's showing error for me. 
Error is following below:
(boxes, scores, classes, num_detections) = ssd_sess.run([boxes, scores, classes, num_detections],feed_dict={image_tensor: image_np_expanded})
  File "C:\python35\lib\site-packages\tensorflow\python\client\session.py", line 900, in run
    run_metadata_ptr)
  File "C:\python35\lib\site-packages\tensorflow\python\client\session.py", line 1078, in _run
    'Cannot interpret feed_dict key as Tensor: ' + e.args[0])
TypeError: Cannot interpret feed_dict key as Tensor: Tensor Tensor("image_tensor:0", shape=(1, ?, ?, 3), dtype=uint8) is not an element of this graph.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by rohitgeo** [comment]
See Prasad's answer at https://stackoverflow.com/questions/50364281/how-to-use-two-models-in-tensorflow-object-detection-api# for the solution for this.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.
